risc0/bigint2/src/field/mod.rs

```diff
@@ -0,0 +1,151 @@
+// operations, but will not work for other use cases (e.g. comparing to a hash value).
+
+pub fn modadd_256_unchecked(
+    lhs: &[u32; FIELD_256_WIDTH_WORDS],
+    rhs: &[u32; FIELD_256_WIDTH_WORDS],
+    modulus: &[u32; FIELD_256_WIDTH_WORDS],
+    result: &mut [u32; FIELD_256_WIDTH_WORDS],
+) {
+    unsafe {
+        sys_bigint2_4(
+            MODADD_256_BLOB.as_ptr(),
+            lhs.as_ptr() as *const u32,
+            rhs.as_ptr() as *const u32,
+            modulus.as_ptr() as *const u32,
+            result.as_mut_ptr() as *mut u32,
+        );
+    }
+}
+
+pub fn modinv_256_unchecked(
+    inp: &[u32; FIELD_256_WIDTH_WORDS],
+    modulus: &[u32; FIELD_256_WIDTH_WORDS],
+    result: &mut [u32; FIELD_256_WIDTH_WORDS],
+) {
+    unsafe {
+        sys_bigint2_3(
+            MODINV_256_BLOB.as_ptr(),
+            inp.as_ptr() as *const u32,
+            modulus.as_ptr() as *const u32,
+            result.as_mut_ptr() as *mut u32,
+        );
+    }
+}
+
+pub fn modmul_256_unchecked(
+    lhs: &[u32; FIELD_256_WIDTH_WORDS],
+    rhs: &[u32; FIELD_256_WIDTH_WORDS],
+    modulus: &[u32; FIELD_256_WIDTH_WORDS],
+    result: &mut [u32; FIELD_256_WIDTH_WORDS],
+) {
+    unsafe {
+        sys_bigint2_4(
+            MODMUL_256_BLOB.as_ptr(),
+            lhs.as_ptr() as *const u32,
+            rhs.as_ptr() as *const u32,
+            modulus.as_ptr() as *const u32,
+            result.as_mut_ptr() as *mut u32,
+        );
+    }
+}
+
+pub fn modsub_256_unchecked(
+    lhs: &[u32; FIELD_256_WIDTH_WORDS],
+    rhs: &[u32; FIELD_256_WIDTH_WORDS],
+    modulus: &[u32; FIELD_256_WIDTH_WORDS],
+    result: &mut [u32; FIELD_256_WIDTH_WORDS],
+) {
+    unsafe {
+        sys_bigint2_4(
+            MODSUB_256_BLOB.as_ptr(),
+            lhs.as_ptr() as *const u32,
+            rhs.as_ptr() as *const u32,
+            modulus.as_ptr() as *const u32,
+            result.as_mut_ptr() as *mut u32,
+        );
+    }
+}
+
+// These "checked" versions verify that `result < modulus`
+
+pub fn modadd_256(
+    lhs: &[u32; FIELD_256_WIDTH_WORDS],
+    rhs: &[u32; FIELD_256_WIDTH_WORDS],
+    modulus: &[u32; FIELD_256_WIDTH_WORDS],
+    result: &mut [u32; FIELD_256_WIDTH_WORDS],
+) {
+    modadd_256_unchecked(&lhs, &rhs, &modulus, result);
+
+    // An honest host will always return a result less than the modulus. A dishonest prover can
+    // sometimes return a result greater than the modulus, so enforce that we're in the honest case.
+    assert!(crate::is_less(&result, &modulus));
+}
+
+pub fn modinv_256(
+    inp: &[u32; FIELD_256_WIDTH_WORDS],
+    modulus: &[u32; FIELD_256_WIDTH_WORDS],
+    result: &mut [u32; FIELD_256_WIDTH_WORDS],
+) {
+    modinv_256_unchecked(&inp, &modulus, result);
+
+    // An honest host will always return a result less than the modulus. A dishonest prover can
+    // sometimes return a result greater than the modulus, so enforce that we're in the honest case.
+    assert!(crate::is_less(&result, &modulus));
+}
```

Comment:
Thing to note about this is that we don't check that `inp` is not zero, so we are in a similar position to the `ec` things where it would panic in host at runtime, which isn't ideal.

Given that we are passing the result buffer as a param, seems the only way to reasonably have this API if handling this is to return a `Result<(), ()>` (or have the error type be a `&'static str` or something minimal) such that those patterns are nudged towards checking

This is optional, just noting that perhaps we should at least document that behaviour